### PR TITLE
[TEP-0135] Refactor CreatePVCsForWorkspaces

### PR DESCRIPTION
**Pull request by QuanZhang-William**
Part of [#6740][#6740] and closes [#6915].

This commit refactors the original `CreatePVCsForWorkspacesWithoutAffinityAssistant` (renamed to `CreatePVCFromVolumeClaimTemplate`) function and its usages to improve readability since the PVC creation logic is now dependent on `AffinityAssistantBehavior`.

There is no functionality change in the commit.

/kind cleanup

[#6740]: #6740
[#6915]: https://github.com/tektoncd/pipeline/issues/6915

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 1]
```diff
@@ -613,18 +613,12 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1.PipelineRun, getPipel
 
 		switch aaBehavior {
 		case affinityassistant.AffinityAssistantPerWorkspace, affinityassistant.AffinityAssistantDisabled:
-			if err = c.createOrUpdateAffinityAssistantsAndPVCs(ctx, pr, aaBehavior); err != nil {
+			failReason, err := c.createOrUpdateAffinityAssistantsAndPVCs(ctx, pr, aaBehavior)
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/pipelinerun.go:616]: Rather than separating the error and the reason for the error, why not use sentinel errors? https://go.dev/blog/go1.13-errors

Here's an example: https://github.com/tektoncd/pipeline/blob/b769b5620300d7fb6d6638083124d03636caa503/pkg/reconciler/pipelinerun/pipelinerun.go#L364-L366

File: pkg/reconciler/pipelinerun/affinity_assistant.go [hunk 2]
```diff
@@ -74,47 +73,49 @@ func (c *Reconciler) createOrUpdateAffinityAssistantsAndPVCs(ctx context.Context
...
-			// so the VolumeClaimTemplates are pass in as PVCs when creating affinity assistant StatefulSet for volume scheduling.
-			// If passed in as VolumeClaimTemplates, the PVCs are owned by Affinity Assistant StatefulSet instead of the pipelinerun.
-			err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []corev1.PersistentVolumeClaimVolumeSource{{ClaimName: claimTemplate.Name}}, unschedulableNodes)
-			errs = append(errs, err...)
+		for claimTemplate, workspace := range claimTemplatesToWorkspace {
+			// To support PVC auto deletion at pipelinerun deletion time, the OwnerReference of the PVCs should be set to the owning pipelinerun,
+			// so we create PVCs from PipelineRuns' VolumeClaimTemplate and pass the PVCs to the Affinity Assistant StatefulSet for volume scheduling.
+			// If passed in as VolumeClaimTemplates directrly, the PVCs are owned by Affinity Assistant StatefulSet instead of the pipelinerun.
+			if err := c.pvcHandler.CreatePVCFromVolumeClaimTemplate(ctx, workspace, *kmeta.NewControllerRef(pr), pr.Namespace); err != nil {
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:92]: can we have pvcHandler.CreatePVCFromVolumeClaimTemplate return the created PVC, and then pass that into createOrUpdateAffinityAssistant rather than building the PersistentVolumeClaimVolumeSource here?

File: pkg/reconciler/pipelinerun/affinity_assistant.go [hunk 3]
```diff
@@ -74,47 +73,49 @@ func (c *Reconciler) createOrUpdateAffinityAssistantsAndPVCs(ctx context.Context
...
 	}
 
 	switch aaBehavior {
 	case aa.AffinityAssistantPerWorkspace:
 		for claim, workspaceName := range claimToWorkspace {
 			aaName := GetAffinityAssistantName(workspaceName, pr.Name)
-			err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []corev1.PersistentVolumeClaimVolumeSource{*claim}, unschedulableNodes)
-			errs = append(errs, err...)
+			if err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []corev1.PersistentVolumeClaimVolumeSource{*claim}, unschedulableNodes); err != nil {
+				return ReasonCouldntCreateOrUpdateAffinityAssistantStatefulSet, errorutils.NewAggregate(err)
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:85]: I don't think it's necessary to use errorutils.NewAggregate for only a single err

File: pkg/reconciler/pipelinerun/affinity_assistant.go [hunk 4]
```diff
@@ -74,47 +73,49 @@ func (c *Reconciler) createOrUpdateAffinityAssistantsAndPVCs(ctx context.Context
...
-			// In AffinityAssistantPerWorkspace mode, the reconciler has created PVCs (owned by pipelinerun) from pipelinerun's VolumeClaimTemplate at this point,
-			// so the VolumeClaimTemplates are pass in as PVCs when creating affinity assistant StatefulSet for volume scheduling.
-			// If passed in as VolumeClaimTemplates, the PVCs are owned by Affinity Assistant StatefulSet instead of the pipelinerun.
-			err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []corev1.PersistentVolumeClaimVolumeSource{{ClaimName: claimTemplate.Name}}, unschedulableNodes)
-			errs = append(errs, err...)
+		for claimTemplate, workspace := range claimTemplatesToWorkspace {
+			// To support PVC auto deletion at pipelinerun deletion time, the OwnerReference of the PVCs should be set to the owning pipelinerun,
+			// so we create PVCs from PipelineRuns' VolumeClaimTemplate and pass the PVCs to the Affinity Assistant StatefulSet for volume scheduling.
+			// If passed in as VolumeClaimTemplates directrly, the PVCs are owned by Affinity Assistant StatefulSet instead of the pipelinerun.
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:91]: I'm confused about this sentence; is this meant as an explanation of why we pass the created PVCs into createOrUpdateAffinityAssistant instead of the volumeclaimtemplates? I wasn't sure if "passed in" refers to user-provided workspace bindings.

File: pkg/reconciler/volumeclaim/pvchandler.go [hunk 5]
```diff
@@ -51,50 +50,52 @@ func NewPVCHandler(clientset clientset.Interface, logger *zap.SugaredLogger) Pvc
...
-		claim.Name = GetPVCNameWithoutAffinityAssistant(workspaceBinding.VolumeClaimTemplate.Name, workspaceBinding, ownerReference)
-		claim.Namespace = namespace
-		claim.OwnerReferences = []metav1.OwnerReference{ownerReference}
-		claims[workspaceBinding.Name] = claim
+		if err == nil {
+			c.logger.Infof("Created PersistentVolumeClaim %s in namespace %s", claim.Name, claim.Namespace)
+		}
+	case err != nil:
+		return fmt.Errorf("failed to retrieve PVC %s: %w", claim.Name, err)
 	}
```
Comment (lbernick) [inline pkg/reconciler/volumeclaim/pvchandler.go:81]: can this conditional logic be simplified?

File: pkg/reconciler/volumeclaim/pvchandler.go [hunk 6]
```diff
@@ -51,50 +50,52 @@ func NewPVCHandler(clientset clientset.Interface, logger *zap.SugaredLogger) Pvc
...
+
+// getPVCFromVolumeClaimTemplate returns a PersistentVolumeClaim based on given workspaceBinding (using VolumeClaimTemplate), ownerReference and namespace
+func (c *defaultPVCHandler) getPVCFromVolumeClaimTemplate(workspaceBinding v1.WorkspaceBinding, ownerReference metav1.OwnerReference, namespace string) *corev1.PersistentVolumeClaim {
+	if workspaceBinding.VolumeClaimTemplate == nil {
+		c.logger.Infof("workspace binding %v does not contain VolumeClaimTemplate, skipping creating PVC", workspaceBinding.Name)
+		return nil
+	}
+
+	claim := workspaceBinding.VolumeClaimTemplate.DeepCopy()
+	claim.Name = GetPVCNameWithoutAffinityAssistant(workspaceBinding.VolumeClaimTemplate.Name, workspaceBinding, ownerReference)
```
Comment (lbernick) [inline pkg/reconciler/volumeclaim/pvchandler.go:94]: maybe a better name for this function could be "GeneratePVCNameFromWorkspaceBinding"?

File: pkg/reconciler/pipelinerun/affinity_assistant.go (hunk 2 above)
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:92]: hmmm, is there any difference? I personally don't find it hard to read.

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:96]: Ohh, I see. I did some investigation and realized all we need  for the `PersistentVolumeClaimVolumeSource` is the claim name. I refactored this section a bit to improve readability.

File: pkg/reconciler/pipelinerun/affinity_assistant.go (hunk 4 above)
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go]: Yeah you are right. I rephrased this comment and the comment below, hoping it is less confusing now.

File: pkg/reconciler/volumeclaim/pvchandler.go (hunk 5 above)
Comment (QuanZhang-William) [reply pkg/reconciler/volumeclaim/pvchandler.go]: I refactored this part a bit. It is not really simplified but I found it is easier to read for me

File: pkg/reconciler/volumeclaim/pvchandler.go (hunk 6 above)
Comment (QuanZhang-William) [reply pkg/reconciler/volumeclaim/pvchandler.go]: SGTM, function name changed

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go [hunk 7]
```diff
@@ -314,6 +315,76 @@ func TestCreateAndDeleteOfAffinityAssistantPerWorkspaceOrDisabled(t *testing.T)
 	}
 }
 
+func TestCreateOrUpdateAffinityAssistantsAndPVCs_Failure(t *testing.T) {
```
Comment (QuanZhang-William) [inline pkg/reconciler/pipelinerun/affinity_assistant_test.go:318]: This test is copied from https://github.com/tektoncd/pipeline/pull/6927

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 8]
```diff
@@ -620,18 +620,20 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1.PipelineRun, getPipel
...
 		switch aaBehavior {
 		case affinityassistant.AffinityAssistantPerWorkspace, affinityassistant.AffinityAssistantDisabled:
-			if err = c.createOrUpdateAffinityAssistantsAndPVCs(ctx, pr, aaBehavior); err != nil {
-				logger.Errorf("Failed to create PVC or affinity assistant StatefulSet for PipelineRun %s: %v", pr.Name, err)
-				if aaBehavior == affinityassistant.AffinityAssistantPerWorkspace {
-					pr.Status.MarkFailed(ReasonCouldntCreateOrUpdateAffinityAssistantStatefulSetPerWorkspace,
-						"Failed to create StatefulSet for PipelineRun %s/%s correctly: %s",
-						pr.Namespace, pr.Name, err)
-				} else {
+			if err := c.createOrUpdateAffinityAssistantsAndPVCs(ctx, pr, aaBehavior); err != nil {
```
Comment (QuanZhang-William) [inline pkg/reconciler/pipelinerun/pipelinerun.go:623]: This part is copied from https://github.com/tektoncd/pipeline/pull/6927

File: pkg/reconciler/pipelinerun/affinity_assistant.go (hunk 3 above)
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go]: Changed to sentinel error

File: pkg/reconciler/pipelinerun/affinity_assistant.go [hunk 9]
```diff
@@ -40,86 +41,92 @@ import (
...
+				return fmt.Errorf("%w: %v", ErrPvcCreationFailed, err) //nolint:errorlint
+			}
+			aaName := GetAffinityAssistantName(workspace.Name, pr.Name)
+			if err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []string{claimTemplate.Name}, unschedulableNodes); err != nil {
+				return fmt.Errorf("%w: %v", ErrAffinityAssistantCreationFailed, err)
+			}
 		}
 	case aa.AffinityAssistantPerPipelineRun, aa.AffinityAssistantPerPipelineRunWithIsolation:
-		if claims != nil || claimTemplates != nil {
+		if claimNames != nil || claimTemplates != nil {
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:105]: nit: I suggest `if len(claimNames) != 0` and likewise for claimTemplates, since what you have here differentiates between a nil slice and empty slice which I don't think is intentional.

File: pkg/reconciler/volumeclaim/pvchandler.go [hunk 10]
```diff
@@ -128,30 +128,29 @@ func (c *defaultPVCHandler) PurgeFinalizerAndDeletePVCForWorkspace(ctx context.C
...
-
-		claim := workspaceBinding.VolumeClaimTemplate.DeepCopy()
-		claim.Name = GetPVCNameWithoutAffinityAssistant(workspaceBinding.VolumeClaimTemplate.Name, workspaceBinding, ownerReference)
-		claim.Namespace = namespace
-		claim.OwnerReferences = []metav1.OwnerReference{ownerReference}
-		claims[workspaceBinding.Name] = claim
+// getPVCFromVolumeClaimTemplate returns a PersistentVolumeClaim based on given workspaceBinding (using VolumeClaimTemplate), ownerReference and namespace
+func (c *defaultPVCHandler) getPVCFromVolumeClaimTemplate(workspaceBinding v1.WorkspaceBinding, ownerReference metav1.OwnerReference, namespace string) *corev1.PersistentVolumeClaim {
+	if workspaceBinding.VolumeClaimTemplate == nil {
+		c.logger.Infof("workspace binding %v does not contain VolumeClaimTemplate, skipping creating PVC", workspaceBinding.Name)
```
Comment (lbernick) [inline pkg/reconciler/volumeclaim/pvchandler.go:134]: nit: this log is misleading since this function doesn't create PVCs; can you rephrase or remove?

**Review by lbernick — APPROVED** [review]

**Comment by QuanZhang-William** [comment]
/hold cancel 

v0.50 LTS is released

File: pkg/reconciler/pipelinerun/affinity_assistant.go (hunk 9 above)
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:105]: Thanks, fixed

File: pkg/reconciler/volumeclaim/pvchandler.go (hunk 10 above)
Comment (QuanZhang-William) [reply pkg/reconciler/volumeclaim/pvchandler.go:134]: Removed

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm
